diff --git a/crates/flux-refineck/src/type_env.rs b/crates/flux-refineck/src/type_env.rs
index f8237f0..ac89964 100644
--- a/crates/flux-refineck/src/type_env.rs
+++ b/crates/flux-refineck/src/type_env.rs
@@ -141,6 +141,15 @@ impl<'a> TypeEnv<'a> {
         let result = self.bindings.lookup_unfolding(infcx, place, span)?;
         if result.is_strg && mutbl == Mutability::Mut {
             Ok(Ty::ptr(PtrKind::Mut(re), result.path()))
+        } else if mutbl == Mutability::Mut && should_infer_mut_borrow_bound(&result.ty) {
+            let t1 = result.ty.clone();
+            let t2 = t1.with_holes().replace_holes(|sorts, kind| {
+                debug_assert_eq!(kind, HoleKind::Pred);
+                infcx.fresh_kvar(sorts, KVarEncoding::Conj)
+            });
+            infcx.subtyping(&t1, &t2, ConstrReason::Other)?;
+            result.block_with(t2.clone());
+            Ok(Ty::mk_ref(re, t2, Mutability::Mut))
         } else {
             // FIXME(nilehmann) we should block the place here. That would require a notion
             // of shared vs mutable block types because sometimes blocked places from a shared
@@ -426,6 +435,13 @@ pub(crate) enum PtrToRefBound {
     Identity,
 }
 
+fn should_infer_mut_borrow_bound(ty: &Ty) -> bool {
+    matches!(
+        ty.as_bty_skipping_existentials(),
+        Some(BaseTy::Adt(adt_def, args)) if adt_def.is_opaque() && args.is_empty()
+    )
+}
+
 impl flux_infer::infer::LocEnv for TypeEnv<'_> {
     fn ptr_to_ref(
         &mut self,
diff --git a/crates/flux-refineck/src/type_env/place_ty.rs b/crates/flux-refineck/src/type_env/place_ty.rs
index f3ab789..189dc5a 100644
--- a/crates/flux-refineck/src/type_env/place_ty.rs
+++ b/crates/flux-refineck/src/type_env/place_ty.rs
@@ -134,11 +134,15 @@ impl PlacesTree {
         while let Some(elem) = cursor.next() {
             match elem {
                 PlaceElem::Deref => {
-                    if let TyKind::Ptr(_, path) = ty.kind() {
-                        cursor.change_root(path);
-                        ty = self.get_loc(&cursor.loc).ty.clone();
-                    } else {
-                        return;
+                    match ty.kind() {
+                        TyKind::Ptr(_, path) => {
+                            cursor.change_root(path);
+                            ty = self.get_loc(&cursor.loc).ty.clone();
+                        }
+                        Ref!(_, deref_ty, _) => {
+                            ty = deref_ty.clone();
+                        }
+                        _ => return,
                     }
                 }
                 PlaceElem::Field(f) => {
