diff --git a/source/vir/src/ast_simplify.rs b/source/vir/src/ast_simplify.rs
index 7ca9b2a9..e11a099b 100644
--- a/source/vir/src/ast_simplify.rs
+++ b/source/vir/src/ast_simplify.rs
@@ -38,7 +38,6 @@ use std::sync::Arc;
 struct ClosureDatatype {
     enclosing_fun: Fun,
     args: Typs,
-    #[allow(dead_code)]
     output: Typ,
     kind: ClosureKind,
     path: Path,
@@ -1347,7 +1346,7 @@ pub fn simplify_krate(ctx: &mut GlobalCtx, krate: &Krate) -> Result<Krate, VirEr
     );
     let functions = vec_map_result(functions, |f| simplify_function(ctx, &mut state, f))?;
     let mut trait_impls = vec_map_result(&trait_impls, |t| simplify_trait_impl(&mut state, t))?;
-    let assoc_type_impls =
+    let mut assoc_type_impls =
         vec_map_result(&assoc_type_impls, |a| simplify_assoc_type_impl(&mut state, a))?;
 
     let functions = vec_map_result(&functions, |f: &Function| {
@@ -1469,12 +1468,13 @@ pub fn simplify_krate(ctx: &mut GlobalCtx, krate: &Krate) -> Result<Krate, VirEr
         datatypes.push(Spanned::new(ctx.no_span.clone(), datatypex));
 
         // Add a trait bound, `ClosureType: {Fn, FnMut, FnOnce}`
-        // TODO: include Output associated type
 
         let typ_args: Typs = Arc::new(
             function.x.typ_params.iter().map(|tb| Arc::new(TypX::TypParam(tb.clone()))).collect(),
         );
-        let self_typ = Arc::new(TypX::Datatype(Dt::Path(closure.path), typ_args, Arc::new(vec![])));
+        let closure_path = closure.path;
+        let self_typ =
+            Arc::new(TypX::Datatype(Dt::Path(closure_path.clone()), typ_args, Arc::new(vec![])));
         let args_tuple_typ =
             Arc::new(TypX::Datatype(Dt::Tuple(closure.args.len()), closure.args, Arc::new(vec![])));
         let impl_path = Arc::new(crate::ast::PathX {
@@ -1486,13 +1486,29 @@ pub fn simplify_krate(ctx: &mut GlobalCtx, krate: &Krate) -> Result<Krate, VirEr
             typ_params: function.x.typ_params.clone(),
             typ_bounds: function.x.typ_bounds.clone(),
             trait_path: closure.kind.trait_path(),
-            trait_typ_args: Arc::new(vec![self_typ, args_tuple_typ]),
+            trait_typ_args: Arc::new(vec![self_typ.clone(), args_tuple_typ.clone()]),
             trait_typ_arg_impls: Spanned::new(ctx.no_span.clone(), Arc::new(vec![])),
             owning_module: None,
             auto_imported: true,
             external_trait_blanket: false,
         };
         trait_impls.push(Spanned::new(ctx.no_span.clone(), trait_implx));
+
+        let assoc_impl_path = Arc::new(crate::ast::PathX {
+            krate: CrateId::Internal,
+            segments: Arc::new(vec![crate::def::impl_closure(ClosureKind::FnOnce, id)]),
+        });
+        let assoc_type_implx = crate::ast::AssocTypeImplX {
+            name: Arc::new("Output".to_string()),
+            impl_path: assoc_impl_path,
+            typ_params: function.x.typ_params.clone(),
+            typ_bounds: function.x.typ_bounds.clone(),
+            trait_path: ClosureKind::FnOnce.trait_path(),
+            trait_typ_args: Arc::new(vec![self_typ, args_tuple_typ]),
+            typ: closure.output,
+            impl_paths: Arc::new(vec![]),
+        };
+        assoc_type_impls.push(Spanned::new(ctx.no_span.clone(), assoc_type_implx));
     }
 
     let traits = traits.clone();
