diff --git a/source/rustc_mir_build/src/builder/mod.rs b/source/rustc_mir_build/src/builder/mod.rs
index 90dcc86..974d914 100644
--- a/source/rustc_mir_build/src/builder/mod.rs
+++ b/source/rustc_mir_build/src/builder/mod.rs
@@ -845,7 +845,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
         for bbdata in self.cfg.basic_blocks.iter_mut() {
             let term = bbdata.terminator_mut();
-            let TerminatorKind::Call { ref mut target, destination, .. } = term.kind else {
+            let TerminatorKind::Call { ref func, ref mut target, destination, .. } = term.kind
+            else {
                 continue;
             };
             let Some(target_bb) = *target else { continue };
@@ -876,7 +877,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     true
                 };
 
-            if !ty_is_inhabited {
+            if !ty_is_inhabited && !crate::verus::is_erased_ghost_value_call_func(func) {
                 // Unreachable code warnings are already emitted during type checking.
                 // However, during type checking, full type information is being
                 // calculated but not yet available, so the check for diverging
diff --git a/source/rustc_mir_build_additional_files/verus.rs b/source/rustc_mir_build_additional_files/verus.rs
index 135c484..7735535 100644
--- a/source/rustc_mir_build_additional_files/verus.rs
+++ b/source/rustc_mir_build_additional_files/verus.rs
@@ -8,7 +8,7 @@ use rustc_hir::def::Res;
 use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_middle::hir::place::{Place, Projection, ProjectionKind};
 use rustc_middle::middle::region;
-use rustc_middle::mir::FakeReadCause;
+use rustc_middle::mir::{Const, FakeReadCause, Operand};
 use rustc_middle::thir;
 use rustc_middle::thir::{
     AdtExprBase, Arm, ArmId, Block, BlockId, BlockSafety, Expr, ExprId, ExprKind, LocalVarId, Pat,
@@ -133,6 +133,23 @@ fn get_verus_erasure_ctxt_option() -> Option<Arc<VerusErasureCtxt>> {
     VERUS_ERASURE_CTXT.read().unwrap().clone()
 }
 
+pub(crate) fn is_erased_ghost_value_call_func<'tcx>(func: &Operand<'tcx>) -> bool {
+    let Some(erasure_ctxt) = get_verus_erasure_ctxt_option() else {
+        return false;
+    };
+
+    match func {
+        Operand::Constant(constant) => match constant.const_ {
+            Const::Val(_, ty) => match ty.kind() {
+                TyKind::FnDef(def_id, _) => *def_id == erasure_ctxt.erased_ghost_value_fn_def_id,
+                _ => false,
+            },
+            _ => false,
+        },
+        _ => false,
+    }
+}
+
 /// Our erasure scheme will fail if this query runs too early, before we initialize the
 /// VerusErasureCtxt. However, there are some items where this happens
 /// intentionally (e.g., for consts, which may be evaluated during type-checking).
