diff --git a/source/rustc_mir_build/src/builder/mod.rs b/source/rustc_mir_build/src/builder/mod.rs
index 90dcc86..713cb5e 100644
--- a/source/rustc_mir_build/src/builder/mod.rs
+++ b/source/rustc_mir_build/src/builder/mod.rs
@@ -845,7 +845,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
         for bbdata in self.cfg.basic_blocks.iter_mut() {
             let term = bbdata.terminator_mut();
-            let TerminatorKind::Call { ref mut target, destination, .. } = term.kind else {
+            let TerminatorKind::Call { ref mut target, destination, ref func, .. } = term.kind
+            else {
                 continue;
             };
             let Some(target_bb) = *target else { continue };
@@ -876,7 +877,14 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     true
                 };
 
-            if !ty_is_inhabited {
+            let is_erased_ghost_value = match func.ty(&self.local_decls, self.tcx).kind() {
+                ty::FnDef(fn_def_id, _) => {
+                    crate::verus::is_erased_ghost_value_fn_def_id(*fn_def_id)
+                }
+                _ => false,
+            };
+
+            if !ty_is_inhabited && !is_erased_ghost_value {
                 // Unreachable code warnings are already emitted during type checking.
                 // However, during type checking, full type information is being
                 // calculated but not yet available, so the check for diverging
diff --git a/source/rustc_mir_build_additional_files/verus.rs b/source/rustc_mir_build_additional_files/verus.rs
index 135c484..e32985d 100644
--- a/source/rustc_mir_build_additional_files/verus.rs
+++ b/source/rustc_mir_build_additional_files/verus.rs
@@ -133,6 +133,11 @@ fn get_verus_erasure_ctxt_option() -> Option<Arc<VerusErasureCtxt>> {
     VERUS_ERASURE_CTXT.read().unwrap().clone()
 }
 
+pub(crate) fn is_erased_ghost_value_fn_def_id(fn_def_id: DefId) -> bool {
+    get_verus_erasure_ctxt_option()
+        .is_some_and(|erasure_ctxt| fn_def_id == erasure_ctxt.erased_ghost_value_fn_def_id)
+}
+
 /// Our erasure scheme will fail if this query runs too early, before we initialize the
 /// VerusErasureCtxt. However, there are some items where this happens
 /// intentionally (e.g., for consts, which may be evaluated during type-checking).
